### PR TITLE
[java] UnitTestShouldIncludeAssert - consider SoftAssertionsExtension

### LINKED ISSUE #4113: [java] False positive from JUnitTestsShouldIncludeAssert with SoftAssertions
**Affects PMD Version:** 

6.48.0

**Rule:**

[JUnitTestsShouldIncludeAssert](https://pmd.sourceforge.io/pmd-6.49.0/pmd_rules_java_bestpractices.html#junittestsshouldincludeassert)

**Description:**

Rule has false positives for some ways of using AssertJ [soft assertions](https://assertj.github.io/doc/#assertj-core-soft-assertions), which call `assertAll` automatically so it's not required to call it explicitly in the tests.

**Code Sample demonstrating the issue:**

```
import org.assertj.core.api.SoftAssertions;
import org.assertj.core.api.junit.jupiter.InjectSoftAssertions;
import org.assertj.core.api.junit.jupiter.SoftAssertionsExtension;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;

@ExtendWith({SoftAssertionsExtension.class})
class ShouldIncludeAssertTest {
  @InjectSoftAssertions private SoftAssertions softly;

  @Test
  void falsePositive1() {
    softly.assertThat(false).isTrue();
    softly.assertThat(true).isFalse();
  }

  @Test
  void falsePositive2() {
    SoftAssertions.assertSoftly(
        softly2 -> {
          softly2.assertThat(false).isTrue();
          softly2.assertThat(true).isFalse();
        });
  }
}
```

**Expected outcome:**

No violations reported, but PMD incorrectly reports a violation for both `falsePositive1` and `falsePositive2`. 

**Running PMD through:** Gradle

### PR DESCRIPTION
**Pull request by adangel**
## Related issues

- Fixes #4113

## Ready?

- [x] Added unit tests for fixed bug/feature
- [x] Passing all unit tests
- [x] Complete build `./mvnw clean verify` passes (checked automatically by github actions)
- [x] Added (in-code) documentation (if needed)

### CODE REVIEW

**Comment by pmd-test** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          1 Message
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="true">Compared to main:<br />
This changeset changes 0 violations,<br />
introduces 0 new violations, 0 new errors and 0 new configuration errors,<br />
removes 0 violations, 0 errors and 0 configuration errors.<br />
<a href="https://github.com/pmd/pmd/actions/runs/11843551312?pr=5334">Download full report as build artifact</a></td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

File: pmd-java/src/main/java/net/sourceforge/pmd/lang/java/rule/bestpractices/UnitTestShouldIncludeAssertRule.java [hunk 1]
```diff
@@ -31,16 +36,29 @@ public UnitTestShouldIncludeAssertRule() {
 
     @Override
     public Object visit(ASTMethodDeclaration method, Object data) {
-        ASTBlock body = method.getBody();
+        boolean usesSoftAssertExtension = usesSoftAssertExtension(method.getEnclosingType());
         Set<String> extraAsserts = getProperty(EXTRA_ASSERT_METHOD_NAMES);
+        Predicate<ASTMethodCall> isAssertCall = usesSoftAssertExtension
+                ? TestFrameworksUtil::isSoftAssert
+                : TestFrameworksUtil::isProbableAssertCall;
```
Comment (jsotuyod) [inline pmd-java/src/main/java/net/sourceforge/pmd/lang/java/rule/bestpractices/UnitTestShouldIncludeAssertRule.java:43]: this assumes they are exclusive, and I wonder if this is accurate… maybe we should replace this to something such as…

```suggestion
        Predicate<ASTMethodCall> isAssertCall = TestFrameworksUtil::isProbableAssertCall;
        if (usesSoftAssertExtension) {
            isAssertCall = isAssertCall.or(TestFrameworksUtil::isSoftAssert);
        }
```

Comment (adangel) [reply pmd-java/src/main/java/net/sourceforge/pmd/lang/java/rule/bestpractices/UnitTestShouldIncludeAssertRule.java:43]: I think I understand your point: One could use the extension to use softassertions, but still use normal assertions.

After reading https://assertj.github.io/doc/#assertj-core-soft-assertions I figured, I misunderstood the usage of [assertSoftly](https://assertj.github.io/doc/#assertj-core-assertSoftly-assertions) - it is an alternative to the JUnit5 extension and implicitly also calls assertAll at the end.

I'll update the PR soon.

**Review by jsotuyod — APPROVED** [review]
Thanks!

**Pull request merged**